src/main/scala/com/scality/clueso/compact/TableFilesCompactor.scala

```diff
@@ -0,0 +1,163 @@
+
+    val dirPath = new Path(path, "_spark_metadata")
+    if (fs.exists(dirPath)) {
+      fs.delete(dirPath, true)
+    }
+  }
+
+  def getSubpartitionsToCompact(landingPartitionPath: String, force: Boolean): Array[Long] = {
+    val subpartitionPaths = fs.listStatus(new Path(landingPartitionPath), new PathFilter {
+      override def accept(path: Path): Boolean = {
+        fs.isDirectory(path) && path.getName.matches(partDirnamePattern.pattern.pattern())
+      }
+    })
+
+    var subpartitionValues = subpartitionPaths
+      .map(_.getPath.getName)
+      .flatMap(subpartDirName => {
+        val regexMatch = partDirnamePattern.findFirstMatchIn(subpartDirName)
+        regexMatch.map { rm => rm.group(2) }
+      })
+      .map(_.toLong)
+
+
+    // force = true – all values
+    // force = false
+    //   subpartitions has 1 value -> List()
+    //   subpartitions has 2+ values -> truncate the biggest one (assumes integer)
+    if (!force) {
+      subpartitionValues = if (subpartitionValues.length > 1) {
+        subpartitionValues.sorted.dropRight(1)
+      } else {
+        // only one subpartition exist, so no compaction will occur
+        Array()
+      }
+    }
+
+    subpartitionValues
+  }
+
+  def removeSubpartitionsFromLanding(landingPartitionPath: String, subPartToCompact: Array[Long]): Unit = {
+    val subpartitionPaths = fs.listStatus(new Path(landingPartitionPath), new PathFilter {
+      override def accept(path: Path): Boolean = {
+        if (fs.isDirectory(path) && path.getName.matches(partDirnamePattern.pattern.pattern())) {
+          val regexMatch = partDirnamePattern.findFirstMatchIn(path.getName)
+          regexMatch.foldLeft(false) { (v, rm) => v || subPartToCompact.contains(rm.group(2).toLong) }
+        } else {
+          false
+        }
+      }
+    })
+
+    subpartitionPaths.foreach(fileStatus => fs.delete(fileStatus.getPath, true))
+  }
+
+  /**
+    * Merges the contents of /landing/<partitionColumnName>=<partitionValue>/`*`/`*`.parquet and deletes
+    * /landing/_spark_metadata. /landing/bucket=mybucketname/ will be composed of multiple subpartitions
+    * with format maxOpIndex=<N>. If only of such subpartitions is existent, the partition bucket=.. won't be
+    * merged into /staging. This can be forced by setting <force> to `true`
+    *
+    * @param partitionColumnName usually equal to 'bucket'
+    * @param partitionValue      usually equal to the name of the bucket
+    * @param numPartitions       num of partitions should be equal or multiple of the number of executors on Query applications
+    * @param force               if true, merges even if there's only one  subpartition maxOpIndex=N
+    */
+  def compactLandingPartition(partitionColumnName: String, partitionValue: String, numPartitions: Int, force: Boolean) = {
+    println(s"Merging partition $partitionColumnName=$partitionValue into $numPartitions files")
+
+    val lockFilePath = lockPath()
+
+    if (acquireLock(lockFilePath)) {
```

Comment:
It returns false if the file exists. However, the file creation on S3 is not sync I think. I can't find any evidence saying one way or the other, but I guess that's just the nature of object stores.
This was being used previously in Alluxio, which I believed to have more guarantees to atomicity (although I found no evidence backing that as well), so I guess this can be vulnerable to race conditions